### PR TITLE
[test, entropy_src] Add entropy_src_edn  chip level test

### LINKED ISSUE #13224: edn - entropy reqs test
# Verify the entropy requests from all peripherals.
- Verify that there are no misconnects between each peripheral requesting entropy.
- Have all entropy sources request entropy in the same test one after to show boot to post boot load.
- Cycling all entropy blocks off and on again.
- Ensure there are no deadlocks and everything works as expected.
- X'ref'ed with each IP test that requsts entropy from EDN.

### PR DESCRIPTION
**Pull request by engdoreis**
# Test
The objective of this test is to have all the peripherals connected to the EDN requesting entropy data.

Fixes #13224

### CODE REVIEW

**Comment by tjaychen** [comment]
@cindychip 
it looks like for `alert_handler` the edn request is dependent on a built-in timer which has a pretty long count (16 bits).
so you'll either need to wait out, or as the design suggests (https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/ip_autogen/alert_handler/rtl/alert_handler.sv#L134) force it to a smaller value.

i think you should at least see one at the beginning of time though? do you not? 

@martin-lueker @mwbranstad 
regarding the FIPS bit, you might need to reinitialize the entropy source. It looks like if you are routing entropy to software in entropy_src the FIPS function is automatically disabled. 

@zi-v @moidx does this sound right to you? Technically for known answer tests we would also want AST in "fips" mode right? Whatever that means?

**Comment by msfschaffner** [comment]
> @msfschaffner do you have any advice on how to force the alert handler to request entropy to edn?

As @tjaychen mentions above, I would also suggest to statically force the `wait_cyc_mask_i` signal to a shorter mask (e.g. `0xF` instead of `0xFFFF`) to increase the rate of entropy requests.

**Comment by mwbranstad** [comment]
I see a comment above with my user name tagged, but now crossed out. Is there an action required here, or does this mean it that no action is needed anymore?

**Comment by tjaychen** [comment]
> I see a comment above with my user name tagged, but now crossed out. Is there an action required here, or does this mean it that no action is needed anymore?

we're good @mwbranstad. I misunderstood the design at first, but after reading it through more closely and checking with @timothytrippel and @moidx i think we are good.

**Comment by mwbranstad** [comment]
thanks @tjaychen for the update.

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (moidx) [inline sw/device/tests/entropy_src_edn_reqs_test.c:88]: Please remove dead code as it will not be tested ;)

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (moidx) [inline sw/device/tests/entropy_src_edn_reqs_test.c:217]: should this be a function instead?

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (moidx) [inline sw/device/tests/entropy_src_edn_reqs_test.c:215]: This is a great test! It would be great if we could run it in a loop multiple times in the FPGA target.

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (moidx) [inline sw/device/tests/entropy_src_edn_reqs_test.c:207]: We should also run the test in continuous mode and configure the EDN to force a CSRNG trigger a couple of times.

Comment (moidx) [reply sw/device/tests/entropy_src_edn_reqs_test.c:207]: Ok to add a TODO linking to an issue if we want to do this later.

File: hw/ip/kmac/data/kmac.hjson
Comment (engdoreis) [reply hw/ip/kmac/data/kmac.hjson:509]: done

File: hw/ip/kmac/data/kmac.hjson
Comment (engdoreis) [reply hw/ip/kmac/data/kmac.hjson:505]: done

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (engdoreis) [reply sw/device/tests/entropy_src_edn_reqs_test.c:45]: done

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (engdoreis) [reply sw/device/tests/entropy_src_edn_reqs_test.c:88]: done

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (engdoreis) [reply sw/device/tests/entropy_src_edn_reqs_test.c:217]: It could be, but this macro has the advantage of logging the right source file in case of error.

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (engdoreis) [reply sw/device/tests/entropy_src_edn_reqs_test.c:215]: What would be the best approach? Maybe check in runtime wether it is running in a FPGA?

File: sw/device/lib/dif/dif_kmac.c
Comment (engdoreis) [reply sw/device/lib/dif/dif_kmac.c]: Fixed

**Comment by engdoreis** [comment]
Thanks folks for the comments, here is an update.

## fips:
   I've rebased to get the new `dif_entropy_src` and even then the fips bit never flips high. Note In the trace below that the ES_ROUTE and ES_TYPE are both disabled.
image

## Alert handler:
>can be overridden in DV and FPV to shorten the wait periods.  

I've never done that before, do you have any example where I can look for reference?

**Comment by tjaychen** [comment]
> Thanks folks for the comments, here is an update.
> 
> ## fips:
> I've rebased to get the new `dif_entropy_src` and even then the fips bit never flips high. Note In the trace below that the ES_ROUTE and ES_TYPE are both disabled. image
> 
> ## Alert handler:
> > can be overridden in DV and FPV to shorten the wait periods.
> 
> I've never done that before, do you have any example where I can look for reference?

let me pull down this PR and have a look later today

**Comment by tjaychen** [comment]
@mwbranstad 
@martin-lueker 
@moidx 
it would be good if you guys can confirm / correct what i'm saying below.

@engdoreis I pulled down your test to have a look, as far as I can tell, here is the issue. 

1. the `edn` module that is distributing to most of the blocks is talking to a `csrng` instance that was instantiated during boot time. Boot time instantiation is done specifically with `non-fips` compliance in mind in order to speed up entropy collection

2. so even though you turn the `fips` bit on later, the `edn0` module still keeps the `fips` bit low because it was instantiated with non-fips compliant entropy.  If you look at `edn1` on the other hand, it talks to a csrng instance that has been instantiated with the right kind of seed.

In terms of what you need to do, i think in the beginning in your test you actually need to disable all of `entropy_src`, `csrng` and `edn` (since it was turned on by `test rom`).  Then proceed with what you have here, THEN I think you will see the fips bit correctly set.

File: sw/device/lib/dif/dif_kmac.c
Comment (engdoreis) [reply sw/device/lib/dif/dif_kmac.c]: You might be right, anyway it seems that I don't need to trigger the reseed to have the kmac requesting to the edn, so I dropped the commit from this PR and I'm going to address that in a future PR.

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (moidx) [reply sw/device/tests/entropy_src_edn_reqs_test.c]: ack

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (moidx) [reply sw/device/tests/entropy_src_edn_reqs_test.c]: That should be ok, maybe based on the return value of `fpga_version()`? I am not sure if we are also planning to add compile time flags for the FPGA.

**Comment by tjaychen** [comment]
> Thanks @tjaychen for the hints, I've enabled the fips in the `rom_test` just in a matter of testing and It acctually worked. However no matter what I try in the test to restart the IPs (edn, csrng and entropy_src) the fips never works in the edn0, just in the edn1. I'm running out of options, do you have any other suggestion? Thanks.

if you can add the latest changes you made (including disabling of the 3 blocks and re-enabling), please push that into this PR and i can have a look.

**Comment by engdoreis** [comment]
@tjaychen I appreciate that! I did the changes in the `sw/device/tests/entropy_src_edn_reqs_test.c` file. It is a little messy because I was trying different scenarios.

**Comment by tjaychen** [comment]
> @tjaychen I appreciate that! I did the changes in the `sw/device/tests/entropy_src_edn_reqs_test.c` file. It is a little messy because I was trying different scenarios.

no worries. I'm debugging a few other things right now, but I will try to get to this afternoon my time.

**Comment by tjaychen** [comment]
@engdoreis can you have a look at #13270 ? 
I noticed a few issues where I have filed separate issues. 
Can you take the hacked up commit and see if it looks as you expect now?

**Comment by tjaychen** [comment]
@engdoreis the fix for edn is in. 
So the test mostly should do what's expected now. The main thing is after disable/enable, you need to wait a while for the entire entropy complex to gather enough "fips compliant" randomness before it will respond to any entropy requests again. 

I've added a fake delay in mine, but you can probably deal with it separately.

The last thing is i would like to speed up the random generation process for the sake of simulation, since it's just so slow otherwise.

**Comment by engdoreis** [comment]
Yes @tjaychen, it works just like it should. What would the next step now? Are you going to create a PR?

**Comment by tjaychen** [comment]
> Yes @tjaychen, it works just like it should. What would the next step now? Are you going to create a PR?

i think we can just continue with this PR. You might just need to add some delay in the test.
Separately I will work with Nuvoton to see if it's okay to add a model option to speed up entropy gathering, but that can be done separately from this test.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_edn_vseq.sv
Comment (engdoreis) [inline hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_edn_vseq.sv:11]: @tjaychen I created this Class to increase the entropy request for the alert_handler. But it seems that Verilator is not using it. Am I missing something?

Comment (tjaychen) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_edn_vseq.sv:11]: all the uvm virtual sequence stuff is not used by verilator. Verilator doesn't have all the capabilities of a commercial simulator so it has its own testbench. You would have to run a vcs / xcelium sim for that to work.

Maybe for now you can throw in a comment on what's supposed to happen and others can take over?

Comment (tjaychen) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_edn_vseq.sv:11]: @engdoreis for a separate test, i also need to shorten the wait cycle mask.
So why don't you not do anything here, I'll put in some kind of common function that we can invoke through simulation arguments.  That should make things easier.

Comment (engdoreis) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_edn_vseq.sv:11]: That sounds good to me @tjaychen

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (engdoreis) [reply sw/device/tests/entropy_src_edn_reqs_test.c]: Done

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (engdoreis) [reply sw/device/tests/entropy_src_edn_reqs_test.c]: done

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_edn_vseq.sv
Comment (tjaychen) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_edn_vseq.sv]: so if you're getting the expected FIPs bits now, the rest lgtm.

**Comment by tjaychen** [comment]
btw @engdoreis , [this PR](https://github.com/lowRISC/opentitan/pull/13355) adds the DV option to shorten the ping activation time. However this will only for DV, it won't work for verilator.

**Comment by engdoreis** [comment]
@tjaychen I tried to execute the test with dvsim but apparently all the tests are failing in the HEAD of master with the error:
```
Offending '(clr_i != set_i)'
```

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (luismarques) [inline sw/device/tests/entropy_src_edn_reqs_test.c:61]:
```suggestion
 * Initialize the aes and start a reseed operation.
```

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (luismarques) [inline sw/device/tests/entropy_src_edn_reqs_test.c:81]: This needs some tweaking but I'm not sure what the correct version should say.

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (luismarques) [inline sw/device/tests/entropy_src_edn_reqs_test.c:104]:
```suggestion
  // For security reasons, the LFSR is periodically reseeded with entropy coming
  // from the EDN. Setting the integrity period to a small value will make the
  // otp_ctrl fetch entropy from the EDN.
```

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (luismarques) [inline sw/device/tests/entropy_src_edn_reqs_test.c:97]:
```suggestion
 * Initialized the opt ctrl and configure it, which is enough to make it request
 * entropy from the edn.
```

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (luismarques) [inline sw/device/tests/entropy_src_edn_reqs_test.c:116]:
```suggestion
 * Initialize the otbn, load the randomness app and start its execution, which
 * will request entropy from the edn.
```

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (luismarques) [inline sw/device/tests/entropy_src_edn_reqs_test.c:145]:
```suggestion
 * Initialize the `rv_core_ibex`, and request 4 words of rnd data.
```

File: sw/device/tests/entropy_src_edn_reqs_test.c
Comment (luismarques) [inline sw/device/tests/entropy_src_edn_reqs_test.c:157]:
```suggestion
 * Initialize the `alert_handler`, configure it up to escalation phase 0.
```

**Review by luismarques — COMMENTED** [review]
I didn't study it deeply but the code seems fine so far. I'm leaving only some suggestions on how to tweak the comments.

File: sw/device/lib/testing/entropy_testutils.c
Comment (engdoreis) [reply sw/device/lib/testing/entropy_testutils.c:76]: done

**Comment by tjaychen** [comment]
> @tjaychen I tried to execute the test with dvsim but apparently all the tests are failing in the HEAD of master with the error:
> 
> ```
> Offending '(clr_i != set_i)'
> ```

@engdoreis are you still seeing dv issues? I haven't seen anything at top of tree, and your CI is also passing.

**Comment by cindychip** [comment]
> @cindychip it looks like for `alert_handler` the edn request is dependent on a built-in timer which has a pretty long count (16 bits). so you'll either need to wait out, or as the design suggests (https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/ip_autogen/alert_handler/rtl/alert_handler.sv#L134) force it to a smaller value.
> 
> i think you should at least see one at the beginning of time though? do you not?
> 
> After looking at the code, and discussing a bit with @timothytrippel / @moidx, the existing fips enable / bypass controls should be fine.
> 
> Specifically for this test, if setting FIPS enable does not actually flip the bit high, it might be that both ES_ROUTE/ES_TYPE bits are set in your test. These two together actually overwrite the FIPS enable, but the current DIFs don't check for that.
> 
> ~@martin-lueker @mwbranstad regarding the FIPS bit, you might need to reinitialize the entropy source. It looks like if you are routing entropy to software in entropy_src the FIPS function is automatically disabled.~
> 
> ~@zi-v @moidx does this sound right to you? Technically for known answer tests we would also want AST in "fips" mode right? Whatever that means?~

Sorry for the late reply, I did add some checkings to make sure EDN requests are fetched.
https://github.com/lowRISC/opentitan/pull/13473

**Comment by engdoreis** [comment]
> > @tjaychen I tried to execute the test with dvsim but apparently all the tests are failing in the HEAD of master with the error:
> > ```
> > Offending '(clr_i != set_i)'
> > ```
> 
> @engdoreis are you still seeing dv issues? I haven't seen anything at top of tree, and your CI is also passing.

Yes, I just rebased and ran `./util/dvsim/dvsim.py hw/top_earlgrey/dv/chip_sim_cfg.hjson -i chip_sw_edn_entropy_reqs`
and this is the output I'm seing.
```
|  Milestone  |      Name      | Tests                    |  Passing  |  Total  |  Pass Rate  |
|:-----------:|:--------------:|:-------------------------|:---------:|:-------:|:-----------:|
|             | Unmapped tests | chip_sw_edn_entropy_reqs |     0     |    1    |   0.00 %    |
|             |                | **TOTAL**                |     0     |    1    |   0.00 %    |

## Failure Buckets

* `Offending '(clr_i != set_i)'` has 1 failures:
    * Test chip_sw_edn_entropy_reqs has 1 failures.
        * 0.chip_sw_edn_entropy_reqs.2501922776\
          Line 370, in log /home/doreis/lowrisc/git/opentitan/scratch/entropy_src_edn_test/chip_earlg
rey_asic-sim-vcs/0.chip_sw_edn_entropy_reqs/out/run.log

                	Offending '(clr_i != set_i)'
                UVM_ERROR @ 4396.240106 us: (prim_count.sv:180) [ASSERT FAILED] SimulClrSet_A
                UVM_INFO @ 4396.240106 us: (uvm_report_catcher.svh:705) [UVM/REPORT/CATCHER]
                --- UVM Report catcher Summary ---
                
INFO: [FlowCfg] [scratch_path]: [chip] [/home/doreis/lowrisc/git/opentitan/scratch/entropy_src_edn_te
st/chip_earlgrey_asic-sim-vcs]
INFO: [FlowCfg] [results:email]: [/home/doreis/lowrisc/git/opentitan/scratch/email.html]
ERROR: [dvsim] Errors were encountered in this run.

          [   legend    ]: [Q: queued, D: dispatched, P: passed, F: failed, K: killed, T: total]     
00:01:58  [    build    ]: [Q: 0, D: 0, P: 1, F: 0, K: 0, T: 1] 100%                                 
00:03:54  [     run     ]: [Q: 0, D: 0, P: 0, F: 1, K: 0, T: 1] 100%    
```

**Comment by engdoreis** [comment]
The issue above has been fixed here #13520

File: sw/device/tests/BUILD
Comment (marnovandermaas) [inline sw/device/tests/BUILD:558]: How is this`randomness` dependency included in the test source?

**Review by marnovandermaas — APPROVED** [review]
LGTM

**Comment by engdoreis** [comment]
@tjaychen and @cindychip, this test is ready, but it depends on #13270 to work.

**Comment by tjaychen** [comment]
> > > @tjaychen and @cindychip, this test is ready, but it depends on the changes made in the rtl here #13270 to work.
> > 
> > 
> > hmm? why do we need #13270? that was mostly an example PR to show case changes. Isn't your original one working now too?
> 
> Maybe I'm missing something, but I tested with verilator and without the change in the RTL `srate_value = 12'd30;` the EDN ack never goes high and the test fails. Is it suppose to work only with dvsim?

ooo i see. hmmm...we did add an option to speed up `srate_value`.  See here (https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/ip/ast/rtl/rng.sv#L67).  Without that change it just takes a LONG time. Do you know if verilator can support that kind of construct? If not, then it's probably better to mark this dvsim only.

**Comment by engdoreis** [comment]
> > > > @tjaychen and @cindychip, this test is ready, but it depends on the changes made in the rtl here #13270 to work.
> > > 
> > > 
> > > hmm? why do we need #13270? that was mostly an example PR to show case changes. Isn't your original one working now too?
> > 
> > 
> > Maybe I'm missing something, but I tested with verilator and without the change in the RTL `srate_value = 12'd30;` the EDN ack never goes high and the test fails. Is it suppose to work only with dvsim?
> 
> ooo i see. hmmm...we did add an option to speed up `srate_value`. See here (https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/ip/ast/rtl/rng.sv#L67). Without that change it just takes a LONG time. Do you know if verilator can support that kind of construct? If not, then it's probably better to mark this dvsim only.

I tested with dvsim and FPGA and the test also fails.

**Comment by tjaychen** [comment]
> > > > > @tjaychen and @cindychip, this test is ready, but it depends on the changes made in the rtl here #13270 to work.
> > > > 
> > > > 
> > > > hmm? why do we need #13270? that was mostly an example PR to show case changes. Isn't your original one working now too?
> > > 
> > > 
> > > Maybe I'm missing something, but I tested with verilator and without the change in the RTL `srate_value = 12'd30;` the EDN ack never goes high and the test fails. Is it suppose to work only with dvsim?
> > 
> > 
> > ooo i see. hmmm...we did add an option to speed up `srate_value`. See here (https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/ip/ast/rtl/rng.sv#L67). Without that change it just takes a LONG time. Do you know if verilator can support that kind of construct? If not, then it's probably better to mark this dvsim only.
> 
> I tested with dvsim and FPGA and the test also fails.

ah okay, let's mark this test for DV only then. The duration shortening won't work for FPGA / verilator, not unless we add a MUCH longer delay in the test themselves.

**Comment by tjaychen** [comment]
@engdoreis can you try this on fpga / verilator now? i think it might work.

**Comment by engdoreis** [comment]
@tjaychen it is passing on verilator, but failing on FPGA. But I haven't investigate.
```
I00000 ottf_main.c:80] Running sw/device/tests/entropy_src_edn_reqs_test.c
I00001 entropy_src_edn_reqs_test.c:209] test_initialize
I00002 entropy_src_edn_reqs_test.c:157] alert_handler_configure
I00003 entropy_src_edn_reqs_test.c:254] Entropy src test 0/5
I00004 entropy_src_edn_reqs_test.c:201] alert_handler_test
I00005 entropy_src_edn_reqs_test.c:66] aes_test
I00006 entropy_src_edn_reqs_test.c:117] otbn_test
I00007 entropy_src_edn_reqs_test.c:133] keymgr_test
I00008 entropy_src_edn_reqs_test.c:98] otp_ctrl_test
I00009 entropy_src_edn_reqs_test.c:85] kmac_test
I00010 entropy_src_edn_reqs_test.c:145] ibex_test
I00011 entropy_src_edn_reqs_test.c:254] Entropy src test 1/5
I00012 entropy_src_edn_reqs_test.c:201] alert_handler_test
I00013 entropy_src_edn_reqs_test.c:66] aes_test
I00014 entropy_src_edn_reqs_test.c:117] otbn_test
I00015 entropy_src_edn_reqs_test.c:98] otp_ctrl_test
I00016 entropy_src_edn_reqs_test.c:85] kmac_test
I00017 entropy_src_edn_reqs_test.c:145] ibex_test
E00018 entropy_src_edn_reqs_test.c:149] DIF-fail: dif_rv_core_ibex_read_rnd_data(&ctx->ibex, &rnd) returns 1
I00019 status.c:34] FAIL!

```

**Comment by tjaychen** [comment]
hmmmm... so the thing about the ibex test is that I feel it's a bit timing
dependent.
Because effectively we are in a race between the software and the
hardware.   What the ibex hardware is supposed to do is that after your
request entropy, it handshakes with the edn hardware to get more entropy.
That "could" end up taking some time (because it needs to go through a drbg
generate loop), and if the read happens to land right in the middle of that
it may not work.

Although having said that..the two ibex tests are pretty far from each
other, and you also have uart logging, so that should be tons of time.
This feels like maybe there's a hidden bug.  I can extend the DV one to
multiple loops also and see what happens.

**Comment by tjaychen** [comment]
@engdoreis i had a look at this, and i think the reason fpga is failing is because the entropy complex depletes its entropy.
This is happening because we didn't place the entropy complex in auto mode. I quickly hacked in something for the purpose of this PR, so let me know if it works for FPGA.

File: sw/device/lib/testing/entropy_testutils.c
Comment (tjaychen) [inline sw/device/lib/testing/entropy_testutils.c:56]: @timothytrippel 
this really needs to be cleaned up still...but i've added it here temporarily for now.

Comment (timothytrippel) [reply sw/device/lib/testing/entropy_testutils.c:56]: no problem, once this merges I will clean up the entropy testutils.

**Comment by tjaychen** [comment]
@engdoreis i finally got around to looking at this.
I think somewhere the entropy setup was broken.  In this test, the healthtest parameters were completely wrong, and the entropy_src was actually doing very nonsensical things, which led to the polling timeout. 

I'm hoping this last fix will address these issues on all platforms.

**Comment by engdoreis** [comment]
@tjaychen Thanks for working on that, I rebased and tested on verilator, vcs and FPGA. It's all working.
Are we ready to merge?

**Comment by tjaychen** [comment]
yes let's merge, thanks for sticking with this. I've approved, but not sure if you want another software person to look through.

**Review by tjaychen — APPROVED** [review]

**Pull request merged**
